Refactor Resolutions

perrygoy opened the pull request:
Resolutions have been the odd-duck in ScreenPy for more than 4 whole versions now. It would be nice if they could fit into the paradigm. 

This PR attempts to bring them into the fold by making Resolutions `Resolvable`, and making them look more like all their `-able` siblings.

perrygoy committed:
checkpoint after one Resolution is refactored.

bandophahita commented on screenpy/actions/see.py:
```diff
@@ -36,11 +36,13 @@ class See:
     question: T_Q
     question_to_log: str
-    resolution: BaseResolution
+    resolution: BaseResolution | Resolvable
```
I can't justify my opinion, but my gut reaction is resolution should have a single type.  I'm a being misguided in that thought?

perrygoy replied:
I think we could just add a `resolve` method to `BaseResolution` that just returns itself, that would probably work. Then `BaseResolution` is also `Resolvable` and all its children would be, too.

bandophahita commented on screenpy/actions/see.py:
```diff
@@ -62,10 +64,17 @@ def perform_as(self: SelfSee, the_actor: Actor) -> None:
         self.question_to_log = get_additive_description(question)
         self.resolution = resolution
-        self.resolution_to_log = resolution.get_line()
+        self.resolution_to_log = get_additive_description(resolution)
```
This worries me a little.  Getting the natural language to work out "well enough" is hard.  I found myself updating the `line` in my custom resolutions just to get them to read nicer.  Using `get_additive_description` means it will pull from the `describe()` method instead of `get_line()`.  We may need to add a default `describe()` to `BaseResolution` or add logic to `get_additive_description` that can utilize `get_line`?

perrygoy replied:
Yes, that was what i was going to do, but i wanted to get your opinion on how Resolutions would change first.

My idea is that `describe` in `BaseResolution` would just return `self.get_line()`, while the New Resolutions would have their own `describe` methods that were a little more flexible.

perrygoy committed:
update pre-commit hooks.

perrygoy committed:
make BaseResolution Describable.

perrygoy committed:
make IsEqualTo Describable.

perrygoy committed:
make ContainsItemMatching Resolvable and Describable.

perrygoy committed:
make ContainsTheEntry Resolvable and Describable.

perrygoy committed:
fix import of Matcher from pyhamcrest.

perrygoy committed:
fix typing for ContainsTheEntry.

perrygoy committed:
make ContainsTheItem Resolvable and Describable.

perrygoy committed:
make ContainsTheText Resolvable and Describable.

perrygoy committed:
make EndsWith Resolvable and Describable.

perrygoy committed:
make HasLength Resolvable and Describable.

perrygoy committed:
make IsCloseTo Resolvable and Describable.

perrygoy committed:
make IsEmpty Resolvable and Describable.

perrygoy committed:
make IsGreaterThanOrEqualTo Resolvable and Describable.

perrygoy committed:
make IsGreaterThan Resolvable and Describable.

perrygoy committed:
fix type hits for IsInBounds.

perrygoy committed:
make IsInRange Resolvable and Describable.

perrygoy committed:
make IsLessThanOrEqualTo Resolvable and Describable.

perrygoy committed:
make IsLessThan Resolvable and Describable.

perrygoy committed:
update get_additive_description for Resolvables.

perrygoy committed:
make IsNot Resolvable and Describable.

perrygoy committed:
make Matches Resolvable and Describable.

perrygoy committed:
make ReadsExactly Resolvable and Describable.

perrygoy committed:
make StartsWith Resolvable and Describable.

perrygoy committed:
fix describe methods' strings.

perrygoy committed:
make ContainsTheKey Resolvable and Describable.

perrygoy committed:
make ContainsTheValue Resolvable and Describable.

perrygoy committed:
remove unneeded self typing.

After re-reading https://peps.python.org/pep-0673/, i realize that
i misunderstood the self typing. We only need to type self when we
are returning the instance from the method!

perrygoy committed:
fix tests after Resolution refactor.

perrygoy committed:
revert type hint to Union for 3.8 and 3.9 support.

perrygoy committed:
fix dictionary union for 3.8 support.

perrygoy committed:
add missing raise. :facepalm:

perrygoy committed:
deprecate BaseResolution.

perrygoy commented:
OK @bandophahita, hit it with a good ol' review. 

The important bits are probably the parts in `base_resolution.py`, `protocols.py`, and `see.py`.

I hope you're as excited as i am about this change! I really like how the Resolutions look now.

bandophahita commented on screenpy/protocols.py:
```diff
@@ -60,18 +76,15 @@ def perform_as(self, the_actor: "Actor") -> None:
 
-    caught_exception: Optional[Exception]
+    def resolve(self) -> Matcher:
+        """Form the Matcher for an assertion.
```
Should resolve return BaseMatcher?

perrygoy replied:
No, because the matchers in PyHamcrest are `Matcher` objects (and `BaseMatcher`, the one external matchers are meant to inherit from, [is also a `Matcher`](https://github.com/hamcrest/PyHamcrest/blob/main/src/hamcrest/core/base_matcher.py#L15)). At least that's what i'm thinking.

If you look at the `is_equal` and all those functions from PyHamcrest, they all return some flavor of `Matcher`.

perrygoy committed:
remove 'Resolvable' check.

bandophahita committed:
Fixing typehint violation

perrygoy committed:
fix grammar in test.

bandophahita committed:
updating typehints

bandophahita committed:
unified annotations!

bandophahita commented:
APPROVE!  github is strict

perrygoy commented:
I approve @bandophahita's work, but we both worked on it so there's no one left to rubber stamp!

I'm just gonna force it through.
